crates/chain/src/local_chain.rs

```diff
@@ -19,215 +170,250 @@ impl ChainOracle for LocalChain {
+    /// all of the same chain.
+    pub fn from_blocks(blocks: BTreeMap<u32, BlockHash>) -> Self {
+        let mut tip: Option<CheckPoint> = None;
+
+        for block in &blocks {
+            match tip {
+                Some(curr) => {
+                    tip = Some(
+                        curr.push(BlockId::from(block))
+                            .expect("BTreeMap is ordered"),
+                    )
+                }
+                None => tip = Some(CheckPoint::new(BlockId::from(block))),
+            }
         }
-    }
 
-    /// Get a reference to a map of block height to hash.
-    pub fn blocks(&self) -> &BTreeMap<u32, BlockHash> {
-        &self.blocks
-    }
+        let chain = Self { index: blocks, tip };
 
-    /// Get the chain tip.
-    pub fn tip(&self) -> Option<BlockId> {
-        self.blocks
-            .iter()
-            .last()
-            .map(|(&height, &hash)| BlockId { height, hash })
+        debug_assert!(chain._check_index_is_consistent_with_tip());
+
+        chain
     }
 
-    /// This is like the sparsechain's logic, expect we must guarantee that all invalidated heights
-    /// are to be re-filled.
-    pub fn determine_changeset(&self, update: &Self) -> Result<ChangeSet, UpdateNotConnectedError> {
-        let update = update.as_ref();
-        let update_tip = match update.keys().last().cloned() {
-            Some(tip) => tip,
-            None => return Ok(ChangeSet::default()),
-        };
+    /// Get the highest checkpoint.
+    pub fn tip(&self) -> Option<CheckPoint> {
+        self.tip.clone()
+    }
 
-        // this is the latest height where both the update and local chain has the same block hash
-        let agreement_height = update
-            .iter()
-            .rev()
-            .find(|&(u_height, u_hash)| self.blocks.get(u_height) == Some(u_hash))
-            .map(|(&height, _)| height);
-
-        // the lower bound of the range to invalidate
-        let invalidate_lb = match agreement_height {
-            Some(height) if height == update_tip => u32::MAX,
-            Some(height) => height + 1,
-            None => 0,
-        };
+    /// Returns whether the [`LocalChain`] is empty (has no checkpoints).
+    pub fn is_empty(&self) -> bool {
+        self.tip.is_none()
+    }
 
-        // the first block's height to invalidate in the local chain
-        let invalidate_from_height = self.blocks.range(invalidate_lb..).next().map(|(&h, _)| h);
+    /// Applies the given `update` to the chain.
+    ///
+    /// The method returns [`ChangeSet`] on success. This represents the applied changes to `self`.
+    ///
+    /// To update, the `update_tip` must *connect* with `self`. If `self` and `update_tip` has a
+    /// mutual checkpoint (same height and hash), it can connect if:
+    /// * The mutual checkpoint is the tip of `self`.
+    /// * An ancestor of `update_tip` has a height which is of the checkpoint one higher than the
```

Comment:
I'm really lost here. Does this mean that if I want to remove a block from `self`, then I have to provide another block with the same height?

Let's say self is `1 -> 2 -> 3 -> 4` and the update is `3 -> 4' -> 5'`. As far as I understand, this is valid (I'm removing block `4` and replacing it with `4'`). Instead, if self was `1 -> 2 -> 3 -> 4` and the update was `3 -> 5'`, this wouldn't be valid, as I'm removing block 4 without giving a replacement. Does this make sense?